### ISSUE TITLE
InvalidMutabilityException when executing Ktor request ion iOS with Kotlin 1.4.0

### ISSUE THREAD

**Issue by jcraane** [Kotlin/kotlinx.coroutines#2215]
When executing a http request in a Kotlin multiplatform iOS app, we receive a InvalidMutabilityException when upgrading to Kotlin 1.4.0. When using 1.4.0-rc everything works fine. See the attached stacktrace for details.

The versions used in the project which fails are:
Kotlin 1.4.0
KTor 1.4.0
coroutines, tried multiple versions (1.3.9, 1.3.9-native-mt, 1.3.8-native-mt-1.4.0-rc)

[stacktrace.txt](https://github.com/Kotlin/kotlinx.coroutines/files/5122162/stacktrace.txt)

The corresponding test project can be found here:
https://github.com/jcraane/KotlinNativeStarter/tree/kotlin_1_4_0 (1.4.0 version that crashes)
https://github.com/jcraane/KotlinNativeStarter/tree/kotlin_1_4rc (1.4.0-rc version that works)
[stacktrace.txt](https://github.com/Kotlin/kotlinx.coroutines/files/5122190/stacktrace.txt)

**Comment by qwwdfsad** [comment]
Hi, I've checked out 1_4_0 version, ran `./gradlew build` and it crashes with `/Users/qwwdfsad/workspace/KotlinNativeStarter/common/src/commonTest/kotlin/sample/SampleTests.kt: (9, 9): Unresolved reference: assertTrue`.

Could you please verify that the project is buildable?

**Comment by jcraane** [comment]
Yes, it indeed fails when executing the ./gradlew build task.

I fixed it (for both the 1.4.0 and 1.4.0-rc branch) and also added some more instructions in
the README.MD to run the project. If you need any more information please let me know.

**Comment by qwwdfsad** [comment]
Stacktrace: 
```
Uncaught Kotlin exception: kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for DispatchedContinuation[MainLoopDispatcher@2ad7ce8, Continuation @ $split$lambda-0$<anonymous>_3COROUTINE$2]. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
    at 0   common                              0x000000010338dd7c kfun:kotlin.Error#<init>(kotlin.String?;kotlin.Throwable?){} + 124 (/Users/teamcity/buildAgent/work/cae0e6559deed4c4/runtime/src/main/kotlin/kotlin/Exceptions.kt:14:63)
    at 1   common                              0x0000000103538783 kfun:kotlinx.coroutines.CoroutinesInternalError#<init>(kotlin.String;kotlin.Throwable){} + 115 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/Exceptions.common.kt:28:77)
    at 2   common                              0x000000010358098a kfun:kotlinx.coroutines.DispatchedTask#handleFatalException(kotlin.Throwable?;kotlin.Throwable?){} + 938 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/internal/DispatchedTask.kt:93:22)
    at 3   common                              0x0000000103580578 kfun:kotlinx.coroutines.DispatchedTask#run(){} + 3512 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/internal/DispatchedTask.kt:64:13)
    at 4   common                              0x00000001032c3a73 kfun:nl.jamiecraane.nativestarter.api.MainLoopDispatcher.dispatch$lambda-0#internal + 195 (/Users/qwwdfsad/workspace/KotlinNativeStarter/common/src/iosX64Main/kotlin/nl.jamiecraane.nativestarter/api/ApplicationDispatcher.kt:17:23)
    at 5   common                              0x00000001032c410e kfun:nl.jamiecraane.nativestarter.api.MainLoopDispatcher.$dispatch$lambda-0$FUNCTION_REFERENCE$3.invoke#internal + 62 (/Users/qwwdfsad/workspace/KotlinNativeStarter/common/src/iosX64Main/kotlin/nl.jamiecraane.nativestarter/api/ApplicationDispatcher.kt:15:51)
    at 6   common                              0x00000001032c416e kfun:nl.jamiecraane.nativestarter.api.MainLoopDispatcher.$dispatch$lambda-0$FUNCTION_REFERENCE$3.$<bridge-UNN>invoke(){}#internal + 62 (/Users/qwwdfsad/workspace/KotlinNativeStarter/common/src/iosX64Main/kotlin/nl.jamiecraane.nativestarter/api/ApplicationDispatcher.kt:15:51)
    at 7   common                              0x00000001032c4587 _4b6f746c696e4e6174697665537461727465723a636f6d6d6f6e_knbridge4 + 183 (/Users/qwwdfsad/workspace/KotlinNativeStarter/common/src/iosX64Main/kotlin/nl.jamiecraane.nativestarter/api/ApplicationDispatcher.kt:15:51)
    at 8   libdispatch.dylib                   0x0000000104b3df11 _dispatch_call_block_and_release + 12
    at 9   libdispatch.dylib                   0x0000000104b3ee8e _dispatch_client_callout + 8
    at 10  libdispatch.dylib                   0x0000000104b4cd97 _dispatch_main_queue_callback_4CF + 1149
    at 11  CoreFoundation                      0x00007fff23da1869 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
    at 12  CoreFoundation                      0x00007fff23d9c3b9 __CFRunLoopRun + 2041
    at 13  CoreFoundation                      0x00007fff23d9b8a4 CFRunLoopRunSpecific + 404
    at 14  GraphicsServices                    0x00007fff38c39bbe GSEventRunModal + 139
    at 15  UIKitCore                           0x00007fff49325968 UIApplicationMain + 1605
    at 16  iosApp                              0x000000010302f948 main + 72 (/Users/qwwdfsad/workspace/KotlinNativeStarter/iosApp/iosApp/AppDelegate.swift:6:7)
    at 17  libdyld.dylib                       0x00007fff520ce1fd start + 1
    at 18  ???                                 0x0000000000000001 0x0 + 1
Caused by: kotlinx.coroutines.CompletionHandlerException: Exception in completion handler ChildCompletion[ChildHandle[DeferredCoroutine{Cancelled}@249fb78], kotlin.Unit] for DeferredCoroutine{Cancelled}@249fb78
    at 0   common                              0x000000010338df3c kfun:kotlin.Exception#<init>(kotlin.String?;kotlin.Throwable?){} + 124 (/Users/teamcity/buildAgent/work/cae0e6559deed4c4/runtime/src/main/kotlin/kotlin/Exceptions.kt:25:63)
    at 1   common                              0x000000010338e18c kfun:kotlin.RuntimeException#<init>(kotlin.String?;kotlin.Throwable?){} + 124 (/Users/teamcity/buildAgent/work/cae0e6559deed4c4/runtime/src/main/kotlin/kotlin/Exceptions.kt:36:63)
    at 2   common                              0x00000001035386f3 kfun:kotlinx.coroutines.CompletionHandlerException#<init>(kotlin.String;kotlin.Throwable){} + 115 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/Exceptions.common.kt:13:78)
    at 3   common                              0x000000010353ddc2 kfun:kotlinx.coroutines.JobSupport.notifyCompletion#internal + 1602 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/JobSupport.kt:365:9)
    at 4   common                              0x000000010353ca06 kfun:kotlinx.coroutines.JobSupport.completeStateFinalization#internal + 1462 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/JobSupport.kt:326:25)
    at 5   common                              0x000000010353abe4 kfun:kotlinx.coroutines.JobSupport.finalizeFinishingState#internal + 1668 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/JobSupport.kt:241:9)
    at 6   common                              0x00000001035477f1 kfun:kotlinx.coroutines.JobSupport.tryMakeCompletingSlowPath#internal + 2001 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/JobSupport.kt:922:16)
    at 7   common                              0x0000000103546efa kfun:kotlinx.coroutines.JobSupport.tryMakeCompleting#internal + 730 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/JobSupport.kt:875:16)
    at 8   common                              0x00000001035468fb kfun:kotlinx.coroutines.JobSupport#makeCompletingOnce(kotlin.Any?){}kotlin.Any? + 603 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/JobSupport.kt:840:30)
    at 9   common                              0x000000010351c714 kfun:kotlinx.coroutines.AbstractCoroutine#resumeWith(kotlin.Result<1:0>){} + 292 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/AbstractCoroutine.kt:111:21)
    at 10  common                              0x00000001033b6e6a kfun:kotlin.coroutines.native.internal.BaseContinuationImpl#resumeWith(kotlin.Result<kotlin.Any?>){} + 1450 (/Users/teamcity/buildAgent/work/cae0e6559deed4c4/runtime/src/main/kotlin/kotlin/coroutines/ContinuationImpl.kt:43:32)
    at 11  common                              0x0000000103580055 kfun:kotlinx.coroutines.DispatchedTask#run(){} + 2197 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/internal/DispatchedTask.kt:53:21)
    at 12  common                              0x00000001032c3a73 kfun:nl.jamiecraane.nativestarter.api.MainLoopDispatcher.dispatch$lambda-0#internal + 195 (/Users/qwwdfsad/workspace/KotlinNativeStarter/common/src/iosX64Main/kotlin/nl.jamiecraane.nativestarter/api/ApplicationDispatcher.kt:17:23)
    ... and 14 more common stack frames skipped
Caused by: kotlin.native.concurrent.InvalidMutabilityException: mutation attempt of frozen kotlinx.coroutines.AwaitAll.AwaitAllNode@927488
    at 0   common                              0x000000010339514d kfun:kotlin.Throwable#<init>(kotlin.String?){} + 93 (/Users/teamcity/buildAgent/work/cae0e6559deed4c4/runtime/src/main/kotlin/kotlin/Throwable.kt:23:37)
    at 1   common                              0x000000010338de8b kfun:kotlin.Exception#<init>(kotlin.String?){} + 91 (/Users/teamcity/buildAgent/work/cae0e6559deed4c4/runtime/src/main/kotlin/kotlin/Exceptions.kt:23:44)
    at 2   common                              0x000000010338e0db kfun:kotlin.RuntimeException#<init>(kotlin.String?){} + 91 (/Users/teamcity/buildAgent/work/cae0e6559deed4c4/runtime/src/main/kotlin/kotlin/Exceptions.kt:34:44)
    at 3   common                              0x00000001033c3e1b kfun:kotlin.native.concurrent.InvalidMutabilityException#<init>(kotlin.String){} + 91 (/Users/teamcity/buildAgent/work/cae0e6559deed4c4/runtime/src/main/kotlin/kotlin/native/concurrent/Freezing.kt:22:60)
    at 4   common                              0x00000001033c56d2 ThrowInvalidMutabilityException + 690 (/Users/teamcity/buildAgent/work/cae0e6559deed4c4/runtime/src/main/kotlin/kotlin/native/concurrent/Internal.kt:92:11)
    at 5   common                              0x00000001034bf3fc MutationCheck + 108
    at 6   common                              0x000000010351ec90 kfun:kotlinx.coroutines.AwaitAll.AwaitAllNode.<set-handle>#internal + 96 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/Await.kt:94:18)
    at 7   common                              0x000000010351de9a kfun:kotlinx.coroutines.AwaitAll.await#internal + 1210 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/Await.kt:68:17)
    at 8   common                              0x000000010351d43d kfun:kotlinx.coroutines.$awaitAllCOROUTINE$1#invokeSuspend(kotlin.Result<kotlin.Any?>){}kotlin.Any? + 1389 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/Await.kt:38:62)
    at 9   common                              0x000000010351d70b kfun:kotlinx.coroutines#awaitAll@kotlin.collections.Collection<kotlinx.coroutines.Deferred<0:0>>(){0§<kotlin.Any?>}kotlin.collections.List<0:0> + 235 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/Await.kt:37:16)
    at 10  common                              0x0000000103689a5f kfun:io.ktor.util.$split$lambda-0COROUTINE$3.invokeSuspend#internal + 2319 (/opt/buildAgent/work/a85294440dc5c6e/ktor-utils/common/src/io/ktor/util/ByteChannels.kt:29:23)
    at 11  common                              0x00000001033b6bb6 kfun:kotlin.coroutines.native.internal.BaseContinuationImpl#resumeWith(kotlin.Result<kotlin.Any?>){} + 758 (/Users/teamcity/buildAgent/work/cae0e6559deed4c4/runtime/src/main/kotlin/kotlin/coroutines/ContinuationImpl.kt:30:39)
    at 12  common                              0x000000010358028f kfun:kotlinx.coroutines.DispatchedTask#run(){} + 2767 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/internal/DispatchedTask.kt:39:50)
    at 13  common                              0x00000001032c3a73 kfun:nl.jamiecraane.nativestarter.api.MainLoopDispatcher.dispatch$lambda-0#internal + 195 (/Users/qwwdfsad/workspace/KotlinNativeStarter/common/src/iosX64Main/kotlin/nl.jamiecraane.nativestarter/api/ApplicationDispatcher.kt:17:23)
    ... and 14 more common stack frames skipped
```

**Comment by qwwdfsad** [comment]
The root cause of this issue seems to be a bug in `kotlinx.coroutines`.
1.4.0 did not regress comparing with 1.4.0-RC, the only change required to fix this is to rollback Ktor version back to 1.3.2-1.4.0-rc from 1.4.0.
Last Ktor release has changed coroutines pattern usage and `AwaitAll` bug started to reproduce on a regular basis

**Comment by jcraane** [comment]
I can indeed confirm that using version 1.3.2-1.4.0-rc of Ktor works! Can we just use this version for now until Ktor fixes this issue for 1.4.0?

**Comment by qwwdfsad** [comment]
Fixed in `1.3.9-native-mt-2`
